### DISCUSSION TITLE
How do I enable and configure per-user keys?

### DISCUSSION THREAD

**Discussion by snaildos** [docker-mailserver/docker-mailserver#4129] [Q&A]
Hey, 
As the title mentions, I'd like to enable per-user keys for encrypting mail. How can this be achieved?
Kind regards,
Snail

**Comment by polarathene** [comment]
You'll get no support regarding this AFAIK.

No maintainer users the feature from when it was contributed, but I did work on a refactor that was going to make that easier. I didn't manage to wrap it up back in 2022 so it's been sitting on my system and needs to be updated to the current state of DMS.

It's on my TODO list after I wrap up LDAP feature work. Unfortunately these tasks take a while for me to budget time for, but provided no disruptions get in the way I'm expecting to have it land by September. We'll have proper tests and documentation once that's ready, but the work did not have an LDAP implementation (_I didn't know LDAP very well back then_), I might be able to support LDAP with the feature once I have looked it over again 👍 

---

For now you'd be on your own. I do know there are past issues (one should still be open), where you should be able to find some technical insights. There's also an old closed PR from 2021-2022 which my work was refactoring.

I recall that per-user keys would have a random encryption key generated, and then encrypt that with the users password hashed to secure it. Thus user would login, their password hash would be input to decrypt the encryption key, and that encryption key could then decrypt the on-disk encrypted mail. I probably explained that a little wrong, since mail should arrive to DMS and get encrypted on-disk for storage, but the "encryption key" (private key?) I mentioned was actually required to decrypt, while a public key allowed encryption but it couldn't decrypt it back.

Alternatively, you might like to try `stalwart` (rust project) instead of DMS. I think that offers LDAP support and on-disk mail encryption too. From what I hear it's quite good, but some still prefer DMS (_one reason apparently is a difference with licenses_).
